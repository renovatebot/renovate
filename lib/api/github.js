const logger = require('winston');
const ghGot = require('gh-got');

const config = {};

module.exports = {
  initRepo,
  // Search
  findFilePaths,
  // Branch
  branchExists,
  getBranchPr,
  // issue
  addAssignees,
  addReviewers,
  addLabels,
  // PR
  findPr,
  checkForClosedPr,
  createPr,
  getPr,
  updatePr,
  // file
  commitFileToBranch,
  getFile,
  getFileContent,
  getFileJson,
};

// Initialize GitHub by getting base branch and SHA
async function initRepo(repoName, token, endpoint) {
  logger.debug(`initRepo(${repoName})`);
  if (token) {
    process.env.GITHUB_TOKEN = token;
  } else if (!process.env.GITHUB_TOKEN) {
    throw new Error(`No token found for GitHub repository ${repoName}`);
  }
  if (endpoint) {
    process.env.GITHUB_ENDPOINT = endpoint;
  }
  config.repoName = repoName;
  try {
    const res = await ghGot(`repos/${repoName}`);
    config.owner = res.body.owner.login;
    logger.debug(`${repoName} owner = ${config.owner}`);
    config.defaultBranch = res.body.default_branch;
    logger.debug(`${repoName} default branch = ${config.defaultBranch}`);
    config.baseCommitSHA = await getBranchCommit(config.defaultBranch);
    config.baseTreeSHA = await getCommitTree(config.baseCommitSHA);
  } catch (err) {
    logger.error(`GitHub init error: ${JSON.stringify(err)}`);
    throw err;
  }
}

// Search

// Returns an array of file paths in current repo matching the fileName
async function findFilePaths(fileName) {
  const res = await ghGot(`search/code?q=repo:${config.repoName}+filename:${fileName}`);
  const exactMatches = res.body.items.filter(item => item.name === fileName);

  // GitHub seems to return files in the root with a leading `/`
  // which then breaks things later on down the line
  return exactMatches.map(item => item.path.replace(/^\//, ''));
}

// Branch

// Returns true if branch exists, otherwise false
async function branchExists(branchName) {
  logger.debug(`Checking if branch exists: ${branchName}`);
  try {
    const res = await ghGot(`repos/${config.repoName}/git/refs/heads/${branchName}`);
    if (res.statusCode === 200) {
      logger.debug('Branch exists');
      return true;
    }
    // This probably shouldn't happen
    logger.debug('Branch doesn\'t exist');
    return false;
  } catch (error) {
    if (error.statusCode === 404) {
      // If file not found, then return false
      logger.debug('Branch doesn\'t exist');
      return false;
    }
    // Propagate if it's any other error
    throw error;
  }
}

// Returns the Pull Request for a branch. Null if not exists.
async function getBranchPr(branchName) {
  logger.debug(`getBranchPr(${branchName})`);
  const gotString = `repos/${config.repoName}/pulls?` +
    `state=open&base=${config.defaultBranch}&head=${config.owner}:${branchName}`;
  const res = await ghGot(gotString);
  if (!res.body.length) {
    return null;
  }
  const prNo = res.body[0].number;
  return getPr(prNo);
}

// Issue

async function addAssignees(issueNo, assignees) {
  logger.debug(`Adding assignees ${assignees} to #${issueNo}`);
  await ghGot.post(`repos/${config.repoName}/issues/${issueNo}/assignees`, {
    body: {
      assignees,
    },
  });
}

async function addReviewers(issueNo, reviewers) {
  logger.debug(`Adding reviewers ${reviewers} to #${issueNo}`);
  await ghGot.post(`repos/${config.repoName}/pulls/${issueNo}/requested_reviewers`, {
    headers: {
      accept: 'application/vnd.github.black-cat-preview+json',
    },
    body: {
      reviewers,
    },
  });
}

async function addLabels(issueNo, labels) {
  logger.debug(`Adding labels ${labels} to #${issueNo}`);
  await ghGot.post(`repos/${config.repoName}/issues/${issueNo}/labels`, {
    body: JSON.stringify(labels),
  });
}

async function findPr(branchName, prTitle, state = 'all') {
  logger.debug(`findPr(${branchName}, ${state})`);
  const urlString = `repos/${config.repoName}/pulls?head=${config.owner}:${branchName}&state=${state}`;
  logger.debug(`findPr urlString: ${urlString}`);
  const res = await ghGot(urlString);
  let pr = null;
  res.body.forEach((result) => {
    if (!prTitle || result.title === prTitle) {
      pr = result;
      if (pr.state === 'closed') {
        pr.isClosed = true;
      }
      pr.displayNumber = `Pull Request #${pr.number}`;
    }
  });
  return pr;
}

// Pull Request
async function checkForClosedPr(branchName, prTitle) {
  logger.debug(`checkForClosedPr(${branchName}, ${prTitle})`);
  const url = `repos/${config.repoName}/pulls?state=closed&head=${config.owner}:${branchName}`;
  const res = await ghGot(url);
  // Return true if any of the titles match exactly
  return res.body.some(pr => pr.title === prTitle && pr.head.label === `${config.owner}:${branchName}`);
}

// Creates PR and returns PR number
async function createPr(branchName, title, body) {
  const pr = (await ghGot.post(`repos/${config.repoName}/pulls`, {
    body: { title, head: branchName, base: config.defaultBranch, body },
  })).body.number;
  pr.displayNumber = `Pull Request #${pr.number}`;
  return pr;
}

// Gets details for a PR
async function getPr(prNo) {
  if (!prNo) {
    return null;
  }
  const pr = (await ghGot(`repos/${config.repoName}/pulls/${prNo}`)).body;
  if (!pr) {
    return null;
  }
  // Harmonise PR values
  pr.displayNumber = `Pull Request #${pr.number}`;
  if (pr.state === 'closed') {
    pr.isClosed = true;
  }
  if (pr.mergeable_state === 'dirty') {
    pr.isUnmergeable = true;
  }
  if (pr.additions * pr.deletions === 1) {
    pr.canRebase = true;
  }
  if (pr.base.sha !== config.baseCommitSHA) {
    pr.isStale = true;
  }
  return pr;
}

async function updatePr(prNo, title, body) {
  await ghGot.patch(`repos/${config.repoName}/pulls/${prNo}`, {
    body: { title, body },
  });
}

// Generic File operations

async function getFile(filePath, branchName = config.defaultBranch) {
  const res = await ghGot(`repos/${config.repoName}/contents/${filePath}?ref=${branchName}`);
  return res.body.content;
}

async function getFileContent(filePath, branchName = config.baseBranch) {
  try {
    const file = await getFile(filePath, branchName);
    return new Buffer(file, 'base64').toString();
  } catch (error) {
    if (error.statusCode === 404) {
      // If file not found, then return null JSON
      return null;
    }
    // Propagate if it's any other error
    throw error;
  }
}

async function getFileJson(filePath, branchName = config.baseBranch) {
  try {
    const file = await getFile(filePath, branchName);
    return JSON.parse(new Buffer(file, 'base64').toString());
  } catch (error) {
    if (error.statusCode === 404) {
      // If file not found, then return null JSON
      return null;
    }
    // Propagate if it's any other error
    throw error;
  }
}

// Add a new commit, create branch if not existing
async function commitFileToBranch(
  branchName,
  fileName,
  fileContents,
  message,
  parentBranch = config.defaultBranch) {
  logger.debug(`commitFileToBrach('${branchName}', '${fileName}', fileContents, message, '${parentBranch})'`);
  const parentCommit = await getBranchCommit(parentBranch);
  const parentTree = await getCommitTree(parentCommit);
  const blob = await createBlob(fileContents);
  const tree = await createTree(parentTree, fileName, blob);
  const commit = await createCommit(parentCommit, tree, message);
  const isBranchExisting = await branchExists(branchName);
  if (isBranchExisting) {
    await updateBranch(branchName, commit);
  } else {
    await createBranch(branchName, commit);
  }
}

// Internal branch operations

// Creates a new branch with provided commit
async function createBranch(branchName, commit = config.baseCommitSHA) {
  await ghGot.post(`repos/${config.repoName}/git/refs`, {
    body: {
      ref: `refs/heads/${branchName}`,
      sha: commit,
    },
  });
}

// Internal: Updates an existing branch to new commit sha
async function updateBranch(branchName, commit) {
  logger.debug(`Updating branch ${branchName} with commit ${commit}`);
  await ghGot.patch(`repos/${config.repoName}/git/refs/heads/${branchName}`, {
    body: {
      sha: commit,
      force: true,
    },
  });
}

// Low-level commit operations

// Create a blob with fileContents and return sha
async function createBlob(fileContents) {
  logger.debug('Creating blob');
  return (await ghGot.post(`repos/${config.repoName}/git/blobs`, {
    body: {
      encoding: 'base64',
      content: new Buffer(fileContents).toString('base64'),
    },
  })).body.sha;
}

// Return the commit SHA for a branch
async function getBranchCommit(branchName) {
  return (await ghGot(`repos/${config.repoName}/git/refs/heads/${branchName}`)).body.object.sha;
}

// Return the tree SHA for a commit
async function getCommitTree(commit) {
  logger.debug(`getCommitTree(${commit})`);
  return (await ghGot(`repos/${config.repoName}/git/commits/${commit}`)).body.tree.sha;
}

// Create a tree and return SHA
async function createTree(baseTree, filePath, fileBlob) {
  logger.debug(`createTree(${baseTree}, ${filePath}, ${fileBlob})`);
  return (await ghGot.post(`repos/${config.repoName}/git/trees`, {
    body: {
      base_tree: baseTree,
      tree: [
        {
          path: filePath,
          mode: '100644',
          type: 'blob',
          sha: fileBlob,
        },
      ],
    },
  })).body.sha;
}

// Create a commit and return commit SHA
async function createCommit(parent, tree, message) {
  logger.debug(`createCommit(${parent}, ${tree}, ${message})`);
  return (await ghGot.post(`repos/${config.repoName}/git/commits`, {
    body: {
      message,
      parents: [parent],
      tree,
    },
  })).body.sha;
}
