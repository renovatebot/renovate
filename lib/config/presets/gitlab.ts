import { api } from '../../platform/gitlab/gl-got-wrapper';
import { logger } from '../../logger';
import { Preset } from './common';

const { get: glGot } = api;

const GitLabApiUrl = 'https://gitlab.com/api/v4/projects';

async function getDefaultBranchName(
  urlEncodedPkgName: string
): Promise<string> {
  const branchesUrl = `${GitLabApiUrl}/${urlEncodedPkgName}/repository/branches`;
  type GlBranch = {
    default: boolean;
    name: string;
  }[];

  const res = await glGot<GlBranch>(branchesUrl);
  const branches = res.body;
  let defautlBranchName = 'master';
  for (const branch of branches) {
    if (branch.default) {
      defautlBranchName = branch.name;
      break;
    }
  }

  return defautlBranchName;
}

export async function getPreset(
  pkgName: string,
  presetName = 'default'
): Promise<Preset> {
  if (presetName !== 'default') {
    // TODO: proper error contructor
    throw new Error(
      // { pkgName, presetName },
      'Sub-preset names are not supported with Gitlab datasource'
    );
  }
  let res: string;
  try {
    const urlEncodedPkgName = encodeURIComponent(pkgName);
    const defautlBranchName = await getDefaultBranchName(urlEncodedPkgName);

    const presetUrl = `${GitLabApiUrl}/${urlEncodedPkgName}/repository/files/renovate.json?ref=${defautlBranchName}`;
    res = Buffer.from(
      (await glGot(presetUrl)).body.content,
      'base64'
    ).toString();
  } catch (err) {
    logger.debug({ err }, 'Failed to retrieve renovate.json from repo');
    throw new Error('dep not found');
  }
  try {
    return JSON.parse(res);
  } catch (err) /* istanbul ignore next */ {
    logger.debug('Failed to parse renovate.json');
    throw new Error('invalid preset JSON');
  }
}
