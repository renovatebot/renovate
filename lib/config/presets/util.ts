import { logger } from '../../logger';
import { ensureTrailingSlash } from '../../util/url';
import type { Preset } from './types';

export const PRESET_DEP_NOT_FOUND = 'dep not found';
export const PRESET_NOT_FOUND = 'preset not found';

export type PresetFetcher = (
  repo: string,
  fileName: string,
  endpoint: string
) => Promise<Preset>;

export type FetchPresetConfig = {
  pkgName: string;
  filePreset: string;
  presetPath?: string;
  endpoint: string;
  fetch: PresetFetcher;
};

export async function fetchPreset({
  pkgName,
  filePreset,
  presetPath,
  endpoint,
  fetch,
}: FetchPresetConfig): Promise<Preset | undefined> {
  // eslint-disable-next-line no-param-reassign
  endpoint = ensureTrailingSlash(endpoint);
  const [fileName, presetName, subPresetName] = filePreset.split('/');
  const pathPrefix = presetPath ? `${presetPath}/` : '';
  const buildFilePath = (name: string): string => `${pathPrefix}${name}`;
  let jsonContent: any | undefined;
  if (fileName === 'default') {
    try {
      jsonContent = await fetch(
        pkgName,
        buildFilePath('default.json'),
        endpoint
      );
    } catch (err) {
      if (err.message !== PRESET_DEP_NOT_FOUND) {
        throw err;
      }
      logger.debug('default.json preset not found - trying renovate.json');
      jsonContent = await fetch(
        pkgName,
        buildFilePath('renovate.json'),
        endpoint
      );
    }
  } else {
    jsonContent = await fetch(
      pkgName,
      buildFilePath(`${fileName}.json`),
      endpoint
    );
  }

  if (!jsonContent) {
    throw new Error(PRESET_DEP_NOT_FOUND);
  }
  if (presetName) {
    const preset = jsonContent[presetName];
    if (!preset) {
      throw new Error(PRESET_NOT_FOUND);
    }
    if (subPresetName) {
      const subPreset = preset[subPresetName];
      if (!subPreset) {
        throw new Error(PRESET_NOT_FOUND);
      }
      return subPreset;
    }
    return preset;
  }
  return jsonContent;
}
