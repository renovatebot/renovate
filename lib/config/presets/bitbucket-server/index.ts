import { logger } from '../../../logger';
import { ExternalHostError } from '../../../types/errors/external-host-error';
import type { FileData } from '../../../types/platform/bitbucket-server';
import {
  BitbucketServerHttp,
  setBaseUrl,
} from '../../../util/http/bitbucket-server';
import type { Preset } from '../types';
import { PRESET_DEP_NOT_FOUND, fetchPreset } from '../util';

const http = new BitbucketServerHttp();

export async function fetchJSONFile(
  repo: string,
  fileName: string,
  endpoint: string
): Promise<Preset> {
  const [projectKey, repositorySlug] = repo.split('/');
  setBaseUrl(endpoint);
  const url = `rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/browse/${fileName}?limit=20000`;
  let res: { body: FileData };
  try {
    res = await http.getJson(url);
  } catch (err) {
    // istanbul ignore if: not testable with nock
    if (err instanceof ExternalHostError) {
      throw err;
    }
    logger.debug(
      { statusCode: err.statusCode, url: `${endpoint}${url}` },
      `Failed to retrieve ${fileName} from repo`
    );
    throw new Error(PRESET_DEP_NOT_FOUND);
  }
  if (!res.body.isLastPage) {
    logger.warn({ size: res.body.size }, 'Renovate config to big');
    throw new Error('invalid preset JSON');
  }
  try {
    const content = res.body.lines.map((l) => l.text).join('');
    const parsed = JSON.parse(content);
    return parsed;
  } catch (err) {
    throw new Error('invalid preset JSON');
  }
}

export function getPresetFromEndpoint(
  pkgName: string,
  filePreset: string,
  presetPath: string,
  endpoint: string
): Promise<Preset> {
  return fetchPreset({
    pkgName,
    filePreset,
    presetPath,
    endpoint,
    fetch: fetchJSONFile,
  });
}
