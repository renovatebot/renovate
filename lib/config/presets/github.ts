import { logger } from '../../logger';
import { Preset } from './common';
import { Http, HttpOptions } from '../../util/http';
import { PLATFORM_FAILURE } from '../../constants/error-messages';

const id = 'github';
const http = new Http(id);

async function fetchJSONFile(repo: string, fileName: string): Promise<Preset> {
  const url = `https://api.github.com/repos/${repo}/contents/${fileName}`;
  const opts: HttpOptions = {
    headers: {
      accept: global.appMode
        ? 'application/vnd.github.machine-man-preview+json'
        : 'application/vnd.github.v3+json',
    },
  };
  let res: { body: { content: string } };
  try {
    res = await http.getJson(url, opts);
  } catch (err) {
    if (err.message === PLATFORM_FAILURE) {
      throw err;
    }
    logger.debug(
      { statusCode: err.statusCodef },
      `Failed to retrieve ${fileName} from repo`
    );
    throw new Error('dep not found');
  }
  try {
    const content = Buffer.from(res.body.content, 'base64').toString();
    const parsed = JSON.parse(content);
    return parsed;
  } catch (err) {
    throw new Error('invalid preset JSON');
  }
}

export async function getPreset(
  pkgName: string,
  presetName = 'default'
): Promise<Preset> {
  if (presetName === 'default') {
    try {
      const defaultJson = await fetchJSONFile(pkgName, 'default.json');
      return defaultJson;
    } catch (err) {
      if (err.message === PLATFORM_FAILURE) {
        throw err;
      }
      if (err.message === 'dep not found') {
        logger.debug('default.json preset not found - trying renovate.json');
        return fetchJSONFile(pkgName, 'renovate.json');
      }
      throw err;
    }
  }
  return fetchJSONFile(pkgName, `${presetName}.json`);
}
