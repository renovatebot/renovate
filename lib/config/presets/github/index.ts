import { PLATFORM_FAILURE } from '../../../constants/error-messages';
import { PLATFORM_TYPE_GITHUB } from '../../../constants/platforms';
import { logger } from '../../../logger';
import { Http, HttpOptions } from '../../../util/http';
import { ensureTrailingSlash } from '../../../util/url';
import { Preset, PresetConfig } from '../common';

const http = new Http(PLATFORM_TYPE_GITHUB);

export async function fetchJSONFile(
  repo: string,
  fileName: string,
  endpoint: string
): Promise<Preset> {
  const url = `${endpoint}repos/${repo}/contents/${fileName}`;
  const opts: HttpOptions = {
    headers: {
      accept: global.appMode
        ? 'application/vnd.github.machine-man-preview+json'
        : 'application/vnd.github.v3+json',
    },
  };
  let res: { body: { content: string } };
  try {
    res = await http.getJson(url, opts);
  } catch (err) {
    if (err.message === PLATFORM_FAILURE) {
      throw err;
    }
    logger.debug(
      { statusCode: err.statusCode },
      `Failed to retrieve ${fileName} from repo`
    );
    throw new Error('dep not found');
  }
  try {
    const content = Buffer.from(res.body.content, 'base64').toString();
    const parsed = JSON.parse(content);
    return parsed;
  } catch (err) {
    throw new Error('invalid preset JSON');
  }
}

export async function getPresetFromEndpoint(
  pkgName: string,
  filePreset: string,
  endpoint = 'https://api.github.com/'
): Promise<Preset> {
  // eslint-disable-next-line no-param-reassign
  endpoint = ensureTrailingSlash(endpoint);
  const [fileName, presetName, subPresetName] = filePreset.split('/');
  let jsonContent: any;
  if (fileName === 'default') {
    try {
      jsonContent = await fetchJSONFile(pkgName, 'default.json', endpoint);
    } catch (err) {
      if (err.message !== 'dep not found') {
        throw err;
      }
      logger.debug('default.json preset not found - trying renovate.json');
      jsonContent = await fetchJSONFile(pkgName, 'renovate.json', endpoint);
    }
  } else {
    jsonContent = await fetchJSONFile(pkgName, `${fileName}.json`, endpoint);
  }
  if (presetName) {
    if (subPresetName) {
      return jsonContent[presetName]
        ? jsonContent[presetName][subPresetName]
        : undefined;
    }
    return jsonContent[presetName];
  }
  return jsonContent;
}

export function getPreset({
  packageName: pkgName,
  presetName = 'default',
}: PresetConfig): Promise<Preset> {
  return getPresetFromEndpoint(pkgName, presetName);
}
