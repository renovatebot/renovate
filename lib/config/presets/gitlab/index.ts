import { logger } from '../../../logger';
import { ExternalHostError } from '../../../types/errors/external-host-error';
import type { GitLabBranch } from '../../../types/platform/gitlab';
import { GitlabHttp } from '../../../util/http/gitlab';
import type { Preset, PresetConfig } from '../types';
import { PRESET_DEP_NOT_FOUND, fetchPreset } from '../util';

const gitlabApi = new GitlabHttp();
export const Endpoint = 'https://gitlab.com/api/v4/';

async function getDefaultBranchName(
  urlEncodedPkgName: string,
  endpoint: string
): Promise<string> {
  const branchesUrl = `${endpoint}projects/${urlEncodedPkgName}/repository/branches`;

  const res = await gitlabApi.getJson<GitLabBranch[]>(branchesUrl);
  const branches = res.body;
  let defaultBranchName = 'master';
  for (const branch of branches) {
    if (branch.default) {
      defaultBranchName = branch.name;
      break;
    }
  }

  return defaultBranchName;
}

export async function fetchJSONFile(
  repo: string,
  fileName: string,
  endpoint: string
): Promise<Preset> {
  let url = endpoint;
  try {
    const urlEncodedRepo = encodeURIComponent(repo);
    const urlEncodedPkgName = encodeURIComponent(fileName);
    const defaultBranchName = await getDefaultBranchName(
      urlEncodedRepo,
      endpoint
    );
    url += `projects/${urlEncodedRepo}/repository/files/${urlEncodedPkgName}/raw?ref=${defaultBranchName}`;
    return (await gitlabApi.getJson<Preset>(url)).body;
  } catch (err) {
    if (err instanceof ExternalHostError) {
      throw err;
    }
    logger.debug(
      { statusCode: err.statusCode, url },
      `Failed to retrieve ${fileName} from repo`
    );
    throw new Error(PRESET_DEP_NOT_FOUND);
  }
}

export function getPresetFromEndpoint(
  pkgName: string,
  presetName: string,
  presetPath: string,
  endpoint = Endpoint
): Promise<Preset> {
  return fetchPreset({
    pkgName,
    filePreset: presetName,
    presetPath,
    endpoint,
    fetch: fetchJSONFile,
  });
}

export function getPreset({
  packageName: pkgName,
  presetPath,
  presetName = 'default',
}: PresetConfig): Promise<Preset> {
  return getPresetFromEndpoint(pkgName, presetName, presetPath, Endpoint);
}
