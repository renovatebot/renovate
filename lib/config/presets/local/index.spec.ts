import { getName, mocked } from '../../../../test/util';
import * as _bitbucketServer from '../bitbucket-server';
import * as _github from '../github';
import * as _gitlab from '../gitlab';
import * as local from '.';

jest.mock('../gitlab');
jest.mock('../github');
jest.mock('../bitbucket-server');

const gitlab = mocked(_gitlab);
const github = mocked(_github);
const bitbucketServer = mocked(_bitbucketServer);

describe(getName(__filename), () => {
  beforeEach(() => {
    jest.resetAllMocks();
    gitlab.getPresetFromEndpoint.mockResolvedValueOnce({ resolved: 'preset' });
    github.getPresetFromEndpoint.mockResolvedValueOnce({ resolved: 'preset' });
  });
  describe('getPreset()', () => {
    it('throws for unsupported platform', async () => {
      await expect(async () => {
        await local.getPreset({
          packageName: 'some/repo',
          presetName: 'default',
          baseConfig: {
            platform: 'unsupported-platform',
          },
        });
      }).rejects.toThrow();
    });
    it('throws for missing platform', async () => {
      await expect(async () => {
        await local.getPreset({
          packageName: 'some/repo',
          presetName: 'default',
          baseConfig: {
            platform: undefined,
          },
        });
      }).rejects.toThrow();
    });
    it('forwards to gitlab', async () => {
      const content = await local.getPreset({
        packageName: 'some/repo',
        presetName: 'default',
        baseConfig: {
          platform: 'GitLab',
        },
      });
      expect(gitlab.getPresetFromEndpoint.mock.calls).toMatchSnapshot();
      expect(content).toMatchSnapshot();
    });
    it('forwards to custom gitlab', async () => {
      const content = await local.getPreset({
        packageName: 'some/repo',
        presetName: 'default',
        baseConfig: {
          platform: 'gitlab',
          endpoint: 'https://gitlab.example.com/api/v4',
        },
      });
      expect(gitlab.getPresetFromEndpoint.mock.calls).toMatchSnapshot();
      expect(content).toMatchSnapshot();
    });

    it('forwards to github', async () => {
      const content = await local.getPreset({
        packageName: 'some/repo',
        baseConfig: {
          platform: 'github',
        },
      });
      expect(github.getPresetFromEndpoint.mock.calls).toMatchSnapshot();
      expect(content).toMatchSnapshot();
    });
    it('forwards to custom github', async () => {
      const content = await local.getPreset({
        packageName: 'some/repo',
        presetName: 'default',
        baseConfig: {
          platform: 'github',
          endpoint: 'https://api.github.example.com',
        },
      });
      expect(github.getPresetFromEndpoint.mock.calls).toMatchSnapshot();
      expect(content).toMatchSnapshot();
    });

    it('forwards to custom bitbucket-server', async () => {
      const content = await local.getPreset({
        packageName: 'some/repo',
        presetName: 'default',
        baseConfig: {
          platform: 'bitbucket-server',
          endpoint: 'https://git.example.com',
        },
      });
      expect(
        bitbucketServer.getPresetFromEndpoint.mock.calls
      ).toMatchSnapshot();
      expect(content).toMatchSnapshot();
    });
  });
});
