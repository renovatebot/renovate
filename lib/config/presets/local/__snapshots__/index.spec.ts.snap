// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config/presets/local getPreset() forwards to custom github 1`] = `
Array [
  Array [
    "some/repo",
    "",
    "https://api.github.example.com",
  ],
]
`;

exports[`config/presets/local getPreset() forwards to custom github 2`] = `
Object {
  "resolved": "preset",
}
`;

exports[`config/presets/local getPreset() forwards to custom gitlab 1`] = `
Array [
  Array [
    "some/repo",
    "",
    "https://gitlab.example.com/api/v4",
  ],
]
`;

exports[`config/presets/local getPreset() forwards to custom gitlab 2`] = `
Object {
  "resolved": "preset",
}
`;

exports[`config/presets/local getPreset() forwards to github 1`] = `
Array [
  Array [
    "some/repo",
    "default",
    undefined,
  ],
]
`;

exports[`config/presets/local getPreset() forwards to github 2`] = `
Object {
  "resolved": "preset",
}
`;

exports[`config/presets/local getPreset() forwards to gitlab 1`] = `
Array [
  Array [
    "some/repo",
    "",
    undefined,
  ],
]
`;

exports[`config/presets/local getPreset() forwards to gitlab 2`] = `
Object {
  "resolved": "preset",
}
`;
